### PR TITLE
Feature/n09 n10

### PR DESCRIPTION
**Pull request by Pietfried**

### CODE REVIEW

File: lib/ocpp/v201/charge_point.cpp [hunk 1]
```diff
@@ -2337,6 +2398,54 @@ void ChargePoint::handle_delete_certificate_req(Call<DeleteCertificateRequest> c
...
+    CustomerInformationResponse response;
+    const auto& msg = call.msg;
+    response.status = CustomerInformationStatusEnum::Accepted;
+
+    if (!msg.report and !msg.clear) {
+        EVLOG_warning << "CSMS sent CustomerInformation.req with both report and clear flags being false";
+        response.status = CustomerInformationStatusEnum::Rejected;
+    }
+
+    if (!msg.customerCertificate.has_value() and !msg.idToken.has_value() and !msg.customerIdentifier.has_value()) {
```
Comment (marcemmers) [inline lib/ocpp/v201/charge_point.cpp:2418]: else if instead? Or is the second if higher priority than the first one?

File: lib/ocpp/v201/charge_point.cpp [hunk 2]
```diff
@@ -2337,6 +2398,54 @@ void ChargePoint::handle_delete_certificate_req(Call<DeleteCertificateRequest> c
...
+            this->clear_customer_information(msg.customerCertificate, msg.idToken, msg.customerIdentifier);
+        }
+
+        const auto max_customer_information_data_length =
+            this->device_model->get_optional_value<bool>(ControllerComponentVariables::MaxCustomerInformationDataLength)
+                .value_or(DEFAULT_MAX_CUSTOMER_INFORMATION_DATA_LENGTH);
+        if (data.length() > max_customer_information_data_length) {
+            EVLOG_warning << "NotifyCustomerInformation.req data field is too large. Cropping it down to: "
+                          << max_customer_information_data_length << "characters";
+            data = data.substr(0, max_customer_information_data_length);
```
Comment (marcemmers) [inline lib/ocpp/v201/charge_point.cpp:2442]: This is a bit of a nitpick but doing an erase would be more efficient here. You are now first creating a copy and then removing the original string.

```suggestion
            data.erase(max_customer_information_data_length);
```

**Review by marcemmers — CHANGES_REQUESTED** [review]

File: lib/ocpp/v201/charge_point.cpp [hunk 3]
```diff
@@ -8,6 +8,8 @@
 
 using namespace std::chrono_literals;
 
+const auto DEFAULT_MAX_CUSTOMER_INFORMATION_DATA_LENGTH = 51200;
```
[the remark this answers is not in the document]
Comment (Pietfried) [reply lib/ocpp/v201/charge_point.cpp:11]: Its an optional variable in the device model in the InternalCtrlr, so if its not present we need a default. Not sure if we want to define this as required

File: lib/ocpp/v201/charge_point.cpp [hunk 4]
```diff
@@ -1436,6 +1473,30 @@ void ChargePoint::notify_event_req(const std::vector<EventData>& events) {
...
+        req.generatedAt = DateTime();
+        req.tbc = data.length() > 512;
+        return req;
+    }();
+
+    ocpp::Call<NotifyCustomerInformationRequest> call(req, this->message_queue->createMessageId());
+    this->send<NotifyCustomerInformationRequest>(call);
+
+    if (data.length() > 512) {
+        // Recursively send next sequence
```
[the remark this answers is not in the document]
Comment (Pietfried) [reply lib/ocpp/v201/charge_point.cpp]: done

**Review by marcemmers — APPROVED** [review]

**Pull request merged**